File: lib/ubuntu/services.ts
import { CommandError, runCommand } from "../command"

export interface UbuntuService {
  unit: string
  name: string
  description: string
  load: string
  active: string
  sub: string
}

export type ServiceAction = "start" | "stop" | "restart" | "reload" | "enable" | "disable"

const SERVICE_UNIT_REGEX = /^[a-zA-Z0-9@._-]+\.service$/

export async function listServices(): Promise<UbuntuService[]> {
  try {
    const { stdout } = await runCommand("systemctl", [
      "list-units",
      "--type=service",
      "--all",
      "--no-pager",
      "--no-legend",
      "--output=json",
    ])

    const units = JSON.parse(stdout) as Array<{
      unit: string
      load: string
      active: string
      sub: string
      description: string
    }>

    return units
      .filter((unit) => unit.unit.endsWith(".service"))
      .map((unit) => ({
        unit: unit.unit,
        name: unit.unit.replace(/\.service$/, ""),
        description: unit.description,
        load: unit.load,
        active: unit.active,
        sub: unit.sub,
      }))
  } catch (error) {
    if (error instanceof CommandError && /invalid argument/.test(error.message)) {
      // Older systemd versions without JSON output support
      const fallback = await runCommand("systemctl", [
        "list-units",
        "--type=service",
        "--all",
        "--no-pager",
        "--no-legend",
      ])
      return parsePlainListUnits(fallback.stdout)
    }
    throw error
  }
}

export async function performServiceAction(unitName: string, action: ServiceAction): Promise<{ message: string }> {
  assertValidServiceUnit(unitName)

  const args = buildServiceArgs(unitName, action)

  try {
    const { stdout, stderr } = await runCommand("systemctl", args, { requireRoot: true })
    return {
      message: stdout || stderr || `systemctl ${action} ${unitName} completed`,
    }
  } catch (error) {
    if (error instanceof CommandError) {
      throw new CommandError(`systemctl ${action} ${unitName} failed`, error.stderr || error.message, error.code)
    }
    throw error
  }
}

function buildServiceArgs(unitName: string, action: ServiceAction): string[] {
  if (action === "enable" || action === "disable") {
    return [action, unitName]
  }
  return [action, unitName]
}

function assertValidServiceUnit(value: string) {
  if (!SERVICE_UNIT_REGEX.test(value)) {
    throw new Error("Invalid service unit name")
  }
}

function parsePlainListUnits(output: string): UbuntuService[] {
  return output
    .split("\n")
    .map((line) => line.trim())
    .filter(Boolean)
    .map((line) => {
      const parts = line.split(/\s+/)
      const unit = parts[0] ?? ""
      const load = parts[1] ?? ""
      const active = parts[2] ?? ""
      const sub = parts[3] ?? ""
      const description = parts.slice(4).join(" ")
      return {
        unit,
        name: unit.replace(/\.service$/, ""),
        description,
        load,
        active,
        sub,
      }
    })
    .filter((service) => SERVICE_UNIT_REGEX.test(service.unit))
}

File: internal/services/services.go
package services

import (
	"fmt"
	"strings"

	"orbit/internal/util"
)

type Service struct {
	Unit        string `json:"unit"`
	Load        string `json:"load"`
	Active      string `json:"active"`
	Sub         string `json:"sub"`
	Description string `json:"description"`
}

func List() ([]Service, error) {
	output, err := util.RunCommand("systemctl", "list-units", "--type=service", "--all", "--no-pager", "--plain", "--no-legend")
	if err != nil {
		return nil, err
	}

	var services []Service
	for _, line := range strings.Split(strings.TrimSpace(output), "\n") {
		if line == "" {
			continue
		}
		fields := strings.Fields(line)
		if len(fields) >= 5 {
			services = append(services, Service{
				Unit:        fields[0],
				Load:        fields[1],
				Active:      fields[2],
				Sub:         fields[3],
				Description: strings.Join(fields[4:], " "),
			})
		}
	}
	return services, nil
}

func Start(unit string) error {
	// Validate unit name to prevent injection
	if !isValidUnitName(unit) {
		return fmt.Errorf("invalid unit name: %s", unit)
	}
	_, err := util.RunCommand("systemctl", "start", unit)
	return err
}

func Stop(unit string) error {
	// Validate unit name to prevent injection
	if !isValidUnitName(unit) {
		return fmt.Errorf("invalid unit name: %s", unit)
	}
	_, err := util.RunCommand("systemctl", "stop", unit)
	return err
}

func Restart(unit string) error {
	// Validate unit name to prevent injection
	if !isValidUnitName(unit) {
		return fmt.Errorf("invalid unit name: %s", unit)
	}
	_, err := util.RunCommand("systemctl", "restart", unit)
	return err
}

func Enable(unit string) error {
	// Validate unit name to prevent injection
	if !isValidUnitName(unit) {
		return fmt.Errorf("invalid unit name: %s", unit)
	}
	_, err := util.RunCommand("systemctl", "enable", unit)
	return err
}

func Disable(unit string) error {
	// Validate unit name to prevent injection
	if !isValidUnitName(unit) {
		return fmt.Errorf("invalid unit name: %s", unit)
	}
	_, err := util.RunCommand("systemctl", "disable", unit)
	return err
}

func GetStatus(unit string) (string, error) {
	// Validate unit name to prevent injection
	if !isValidUnitName(unit) {
		return "", fmt.Errorf("invalid unit name: %s", unit)
	}
	output, err := util.RunCommand("systemctl", "status", unit)
	return output, err
}

func GetLogs(unit string, lines int) (string, error) {
	// Validate unit name to prevent injection
	if !isValidUnitName(unit) {
		return "", fmt.Errorf("invalid unit name: %s", unit)
	}
	
	// BUG FIX: Proper int to string conversion
	linesStr := "50"
	if lines > 0 {
		linesStr = fmt.Sprintf("%d", lines)
	}
	output, err := util.RunCommand("journalctl", "-u", unit, "-n", linesStr, "--no-pager")
	return output, err
}

// isValidUnitName validates systemd unit names to prevent command injection
func isValidUnitName(unit string) bool {
	if unit == "" || len(unit) > 256 {
		return false
	}
	// Allow alphanumeric, dash, underscore, dot, @, colon
	// Typical systemd unit: service.name, service@instance.service
	for _, c := range unit {
		if !((c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || 
			(c >= '0' && c <= '9') || c == '-' || c == '_' || 
			c == '.' || c == '@' || c == ':') {
			return false
		}
	}
	return true
}

